commit d01323cff5fe9ed4457898dc5dbea7b31d91ff5a
Author: Tomasz Figa <tomasz.figa@gmail.com>
Date:   Tue Sep 7 18:42:10 2010 +0200

    Backported PMEM from Android GIT kernel.

diff --git a/drivers/misc/pmem.c b/drivers/misc/pmem.c
index 84422f1..fd8992c 100644
--- a/drivers/misc/pmem.c
+++ b/drivers/misc/pmem.c
@@ -594,8 +594,7 @@ static int pmem_mmap(struct file *file, struct vm_area_struct *vma)
 	down_write(&data->sem);
 	/* check this file isn't already mmaped, for submaps check this file
 	 * has never been mmaped */
-	if ((data->flags & PMEM_FLAGS_MASTERMAP) ||
-	    (data->flags & PMEM_FLAGS_SUBMAP) ||
+	if ((data->flags & PMEM_FLAGS_SUBMAP) ||
 	    (data->flags & PMEM_FLAGS_UNSUBMAP)) {
 #if PMEM_DEBUG
 		printk(KERN_ERR "pmem: you can only mmap a pmem file once, "
@@ -757,6 +756,7 @@ end:
 	fput(file);
 	return -1;
 }
+EXPORT_SYMBOL(get_pmem_file);
 
 void put_pmem_file(struct file *file)
 {
@@ -779,6 +779,7 @@ void put_pmem_file(struct file *file)
 #endif
 	fput(file);
 }
+EXPORT_SYMBOL(put_pmem_file);
 
 void flush_pmem_file(struct file *file, unsigned long offset, unsigned long len)
 {
@@ -820,6 +821,7 @@ void flush_pmem_file(struct file *file, unsigned long offset, unsigned long len)
 end:
 	up_read(&data->sem);
 }
+EXPORT_SYMBOL(flush_pmem_file);
 
 static int pmem_connect(unsigned long connect, struct file *file)
 {
@@ -1148,6 +1150,16 @@ static long pmem_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 		DLOG("connect\n");
 		return pmem_connect(arg, file);
 		break;
+	case PMEM_CACHE_FLUSH:
+		{
+			struct pmem_region region;
+			DLOG("flush\n");
+			if (copy_from_user(&region, (void __user *)arg,
+					   sizeof(struct pmem_region)))
+				return -EFAULT;
+			flush_pmem_file(file, region.offset, region.len);
+			break;
+		}
 	default:
 		if (pmem[id].ioctl)
 			return pmem[id].ioctl(file, cmd, arg);
diff --git a/include/linux/android_pmem.h b/include/linux/android_pmem.h
index 398997c..f633621 100644
--- a/include/linux/android_pmem.h
+++ b/include/linux/android_pmem.h
@@ -33,6 +33,7 @@
  * struct (with offset set to 0). 
  */
 #define PMEM_GET_TOTAL_SIZE	_IOW(PMEM_IOCTL_MAGIC, 7, unsigned int)
+#define PMEM_CACHE_FLUSH	_IOW(PMEM_IOCTL_MAGIC, 8, unsigned int)
 
 struct android_pmem_platform_data
 {
